tfjs-core/src/ops/draw_test.ts

```diff
@@ -16,89 +16,206 @@
+
+      if (dtype === 'float32') {
+        if (value < 0 || value > 1) {
+          throw new Error(
+              `Tensor values for a float32 Tensor must be in the ` +
+              `range [0 - 1] but encountered ${value}.`);
+        }
+      } else if (dtype === 'int32') {
+        if (value < 0 || value > 255) {
+          throw new Error(
+              `Tensor values for a int32 Tensor must be in the ` +
+              `range [0 - 255] but encountered ${value}.`);
+        }
+      }
+
+      if (depth === 1) {
+        rgba[0] = value * multiplier;
+        rgba[1] = value * multiplier;
+        rgba[2] = value * multiplier;
+      } else {
+        rgba[d] = value * multiplier;
+      }
     }
-    return this.context;
+
+    const j = i * 4;
+    bytes[j + 0] = Math.round(rgba[0]);
+    bytes[j + 1] = Math.round(rgba[1]);
+    bytes[j + 2] = Math.round(rgba[2]);
+    bytes[j + 3] = Math.round(rgba[3]);
   }
+  return bytes;
 }
 
-describeWithFlags('Draw on 2d context', BROWSER_ENVS, () => {
+async function drawAndReadback(
+    contextType: string, data: number[], shape: number[], dtype: string,
+    alpha = 1) {
+  const [width, height] = shape.slice(0, 2);
+  let img;
+  if (shape.length === 3) {
+    img = tf.tensor3d(
+        data, shape as [number, number, number], dtype as keyof tf.DataTypeMap);
+  } else {
+    img = tf.tensor2d(
+        data, shape as [number, number], dtype as keyof tf.DataTypeMap);
+  }
+  const canvas = getCanvas();
+  const drawOptions = {contextOptions: {contextType}, imageOptions: {alpha}};
+  // tslint:disable-next-line:no-any
+  tf.browser.draw(img, canvas as any, drawOptions);
+  const actualData =
+      await readPixelsFromCanvas(canvas, contextType, width, height);
+  const expectedData = convertToRGBA(data, shape, dtype, alpha);
+  img.dispose();
+  return [actualData, expectedData];
+}
+
+describeWithFlags('draw on canvas context', BROWSER_CPU_WEBGPU_ENVS, (env) => {
+  let contextType: string;
+  beforeAll(async () => {
+    await tf.setBackend(env.name);
+    contextType = env.name === 'cpu' ? '2d' : env.name;
+  });
+
   it('draw image with 4 channels and int values', async () => {
-    const data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16];
-    const img = tf.tensor3d(data, [2, 2, 4], 'int32');
-    const canvas = new MockCanvas(2, 2);
-    const ctx = canvas.getContext('2d');
-
-    // tslint:disable-next-line:no-any
-    tf.browser.draw(img, canvas as any, {contextOptions: {contextType: '2d'}});
-    expectArraysEqual(ctx.getImageData().data, data);
+    const data =
+        [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160];
+    const shape = [2, 2, 4];
+    const dtype = 'int32';
+    const startNumTensors = tf.memory().numTensors;
+    const [actualData, expectedData] =
+        await drawAndReadback(contextType, data, shape, dtype);
+    expect(tf.memory().numTensors).toEqual(startNumTensors);
+    expectArraysClose(actualData, expectedData, 3);
```

Comment:
How do you decide the epsilon value?

Reply:
Add DRAW_EPSILON  and comments about this:
// CPU and GPU handle pixel value differently. The epsilon may possibly grow
// after each draw and read back. The empirical value is 3.0.
const DRAW_EPSILON = 3.0;